File: merge_sort/go/merge_sort_test.go
package merge_sort

import (
	"math/rand"
	"reflect"
	"sort"
	"testing"
	"time"
)

func TestMergeSort(t *testing.T) {
	expected := []int{10, 20, 30, 40, 50, 60, 70}
	unsorted := []int{70, 50, 30, 10, 20, 40, 60}
	sorted := MergeSort(unsorted)
	if !reflect.DeepEqual(sorted, expected) {
		t.Errorf("Sorting is incorrect, got: %v, want: %v", sorted, expected)
	}
}

func TestMergeSort_EmptyArray(t *testing.T) {
	var empty []int
	sorted := MergeSort(empty)
	if len(sorted) != 0 {
		t.Error("Expected empty array, got:", sorted)
	}
}

func TestMergeSort_SingleElementArray(t *testing.T) {
	input := []int{42}
	sorted := MergeSort(input)
	if !reflect.DeepEqual(sorted, input) {
		t.Error("Expected:", input, "got:", sorted)
	}
}

func TestMergeSort_AlreadySortedArray(t *testing.T) {
	input := []int{10, 20, 30, 40, 50, 60, 70}
	sorted := MergeSort(input)
	if !reflect.DeepEqual(sorted, input) {
		t.Error("Expected:", input, "got:", sorted)
	}
}

func TestMergeSort_ReverseSortedArray(t *testing.T) {
	input := []int{70, 60, 50, 40, 30, 20, 10}
	expected := []int{10, 20, 30, 40, 50, 60, 70}
	sorted := MergeSort(input)
	if !reflect.DeepEqual(sorted, expected) {
		t.Error("Expected:", expected, "got:", sorted)
	}
}

func TestMergeSort_DuplicateElementsArray(t *testing.T) {
	input := []int{30, 50, 10, 20, 10, 30, 50}
	expected := []int{10, 10, 20, 30, 30, 50, 50}
	sorted := MergeSort(input)
	if !reflect.DeepEqual(sorted, expected) {
		t.Error("Expected:", expected, "got:", sorted)
	}
}

func TestMergeSort_BigArray(t *testing.T) {
	rand.New(rand.NewSource(time.Now().UnixNano()))

	const size = 1000000
	input := make([]int, size)
	for i := range input {
		input[i] = rand.Intn(size * 10)
	}

	expected := make([]int, len(input))
	copy(expected, input)
	sort.Ints(expected)
	sorted := MergeSort(input)
	if !reflect.DeepEqual(sorted, expected) {
		t.Error("Sorting big array failed")
	}
}

File: merge_sort/python/merge_sort_test.py
import unittest
import random
import time
import copy
from merge_sort import merge_sort


class TestMergeSort(unittest.TestCase):

    def test_merge_sort(self):
        expected = [10, 20, 30, 40, 50, 60, 70]
        unsorted = [70, 50, 30, 10, 20, 40, 60]
        sorted_array = merge_sort(unsorted)
        self.assertEqual(sorted_array, expected)

    def test_merge_sort_empty_array(self):
        empty = []
        sorted_array = merge_sort(empty)
        self.assertEqual(sorted_array, [])

    def test_merge_sort_single_element_array(self):
        input_array = [42]
        sorted_array = merge_sort(input_array)
        self.assertEqual(sorted_array, input_array)

    def test_merge_sort_already_sorted_array(self):
        input_array = [10, 20, 30, 40, 50, 60, 70]
        sorted_array = merge_sort(input_array)
        self.assertEqual(sorted_array, input_array)

    def test_merge_sort_reverse_sorted_array(self):
        input_array = [70, 60, 50, 40, 30, 20, 10]
        expected = [10, 20, 30, 40, 50, 60, 70]
        sorted_array = merge_sort(input_array)
        self.assertEqual(sorted_array, expected)

    def test_merge_sort_duplicate_elements_array(self):
        input_array = [30, 50, 10, 20, 10, 30, 50]
        expected = [10, 10, 20, 30, 30, 50, 50]
        sorted_array = merge_sort(input_array)
        self.assertEqual(sorted_array, expected)

    def test_merge_sort_big_array(self):
        random.seed(time.time())
        size = 100000
        input_array = [random.randint(0, size * 10) for _ in range(size)]
        expected = sorted(copy.deepcopy(input_array))
        sorted_array = merge_sort(input_array)
        self.assertEqual(sorted_array, expected)


if __name__ == "__main__":
    unittest.main()
